tools/codegen/gen.py

```diff
@@ -243,13 +248,111 @@ def __call__(self, f: Union[StructuredNativeFunctions, NativeFunction]) -> List[
         else:
             assert_never(f)
 
+    def gen_structured_class_set_output(self, k: SchemaKind, parent_class: str, generate_super: bool) -> str:
+        if generate_super:
+            set_output_super = f"{parent_class}::set_output(output_idx, sizes, strides, options, names);"
+        else:
+            set_output_super = ""
+        return f"""
+void set_output(int64_t output_idx, IntArrayRef sizes, IntArrayRef strides,
+                TensorOptions options, DimnameList names) override {{
+    {self.gen_structured_class_set_output_body(k)}
+    if (!names.empty()) namedinference::propagate_names(outputs_[output_idx], names);
+    // super must happen after, so that downstream can use maybe_get_output
+    // to retrieve the output
+    {set_output_super}
+}}
+"""
+
+    def gen_structured_class_set_output_body(self, k: SchemaKind) -> str:
+        if k is SchemaKind.functional:
+            if self.dispatch_key == "Meta":
+                return """
+if (strides.empty()) {
+    outputs_[output_idx] = at::empty_meta(sizes, options);
+} else {
+    TORCH_INTERNAL_ASSERT(0, "not implemented yet");
+}
+"""
+            else:
+                expanded_topts = "optTypeMetaToScalarType(options.dtype_opt()), options.layout_opt(), " \
+                    "options.device_opt(), options.pinned_memory_opt()"
+                if self.dispatch_key == "CPU":
+                    empty_impl = "at::native::empty_cpu"
+                    empty_strided_impl = "at::native::empty_strided_cpu"
+                elif self.dispatch_key == "CUDA":
+                    empty_impl = "at::native::empty_cuda"
```

Comment:
Does codegen handle a situation of incorrect device guard (happens on ops with scalars)?

Reply:
I didn't write any code to explicitly do this, so it seems likely that we may have gotten it wrong here. (Maybe TensorIterator is taking care of it, so we don't have any other cases where something wrong happens?)

Reply:
I don't think TI does, when I tried replacing at::empty in TI with at::native::empty I had to manually put deviceGuard there to pacify the tests. I think cpu scalar - cuda tensor interop tests are only for `mul` so you are not hitting them with `add`

Reply:
Yes, that was exactly right. I ported mul and triggered this test failure:

```
============================================================= FAILURES =============================================================
________________________________ TestBinaryUfuncsCUDA.test_binary_op_scalar_device_unspecified_cuda ________________________________

self = <test_binary_ufuncs.TestBinaryUfuncsCUDA testMethod=test_binary_op_scalar_device_unspecified_cuda>
devices = ['cuda:0', 'cuda:1']

    @deviceCountAtLeast(2)
    @onlyCUDA
    def test_binary_op_scalar_device_unspecified(self, devices):
        scalar_val = torch.tensor(1.)
        for default_device in devices:
            with torch.cuda.device(default_device):
                for device in devices:
                    device_obj = torch.device(device)
                    x = torch.rand(3, device=device)
                    y0 = x * scalar_val
                    self.assertEqual(y0.device, device_obj)
                    y1 = scalar_val * x
>                   self.assertEqual(y1.device, device_obj)

test/test_binary_ufuncs.py:715: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
torch/testing/_internal/common_utils.py:1194: in assertEqual
    super().assertEqual(x, y, msg=msg)
E   AssertionError: device(type='cuda', index=0) != device(type='cuda', index=1)
```

Reply:
I came up with a very clever solution for doing device guards that let me delete the device guard generating code entirely from structured kernels, just pushed it here.